File: DAA/Java/fractional_knapsack_float.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

class Main {
    static class Item {
        int profit;
        int weight;

        public Item(int profit, int weight) {
            this.profit = profit;
            this.weight = weight;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter Number of Objects: ");
        int n = scanner.nextInt();

        System.out.println("Enter Profit and Weight (P W):");
        Item[] items = new Item[n];
        for (int i = 0; i < n; i++) {
            int profit = scanner.nextInt();
            int weight = scanner.nextInt();
            items[i] = new Item(profit, weight);
        }

        System.out.print("Enter the Capacity: ");
        int w = scanner.nextInt();

        long startTime = System.nanoTime();

        Arrays.sort(items, new Comparator<Item>() {
            @Override
            public int compare(Item item1, Item item2) {
                double v1 = (double) item1.profit / item1.weight;
                double v2 = (double) item2.profit / item2.weight;
                return Double.compare(v2, v1);
            }
        });

        double ans = 0;
        double[] solution = new double[n];

        for (int i = 0; i < n; i++) {
            if (w >= items[i].weight) {
                ans += items[i].profit;
                w -= items[i].weight;
                solution[i] = 1.0;
            } else {
                double vw = (double) items[i].profit / items[i].weight;
                ans += vw * w;
                solution[i] = (double) w / items[i].weight;
                w = 0;
                break;
            }
        }

        long endTime = System.nanoTime();
        long duration = (endTime - startTime) / 1_000_000; // Convert to milliseconds

        System.out.printf("Optimal Solution Vector: (%s)\n", Arrays.toString(solution));
        System.out.printf("Optimal Value: %.2f\n", ans);
        System.out.printf("Execution Time: %d milliseconds\n", duration);
    }
}

File: daa_python/fractional_knapsack_float.py
def compare(p1, p2):
    v1 = p1[0] / p1[1]
    v2 = p2[0] / p2[1]
    return v1 > v2

n = int(input("Enter Number of Objects: "))
print("Enter Profit and Weight (P W):")
a = []
for i in range(n):
    p, w = map(int, input().split())
    a.append((p, w))

w = int(input("Enter the Capacity: "))

# Start timing
import time
start_time = time.time()

a.sort(key=compare, reverse=True)

solution = [0.0] * n  # Solution list to track selected objects (float data type).

ans = 0
for i in range(n):
    if w >= a[i][1]:
        ans += a[i][0]
        w -= a[i][1]
        solution[i] = 1.0  # Mark the object as selected in the solution list.
        continue
    vw = a[i][0] / a[i][1]
    ans += vw * w
    solution[i] = w / a[i][1]  # Store the fraction as a float.
    w = 0
    break

# Stop timing
end_time = time.time()

print("Optimal Solution Vector: (", end="")
for i in range(n):
    print(f"{solution[i]:.2f}", end=" ")
print(")")

print(f"Optimal Value: {ans:.2f}")

execution_time = (end_time - start_time) * 1000  # Convert to milliseconds
print(f"Execution Time: {execution_time:.2f} milliseconds")
'''Enter Number of Objects: 4
Enter Profit and Weight (P W):
5 2
3 1
8 5
2 3
Enter the Capacity: 5
Optimal Solution Vector: (1.00 1.00 1.00 0.33 )
Optimal Value: 17.33
Execution Time: 0.12 milliseconds
'''